src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,457 @@
+      List<String> result = new ArrayList<>(target);
+      deltas.sort(DeltaComparator.INSTANCE);
+      ListIterator<Delta<String>> it = deltas.listIterator(deltas.size());
+
+      while(it.hasPrevious()) {
+        Delta<String> delta = it.previous();
+        applyTo(delta, result);
+      }
+
+      return result;
+    }
+
+    /**
+     * This function first tries to apply the Delta without any offset, if that fails, then
+     * it tries to apply the Delta with an offset, starting from 1, up to the total lines in
+     * the original content. For every offset, we try both forwards and backwards.
+     */
+    private static void applyTo(Delta<String> delta, List<String> result)
+        throws PatchFailedException {
+      PatchFailedException e = applyDelta(delta, result);
+      if (e == null) {
+        return;
+      }
+
+      Chunk<String> original = delta.getOriginal();
+      Chunk<String> revised = delta.getRevised();
+      int[] direction = {1, -1};
+      int maxOffset = result.size();
+      for (int i = 1; i < maxOffset; i++) {
+        for (int j = 0; j < 2; j++) {
+          int offset = i * direction[j];
+          if (offset + original.getPosition() < 0 || offset + revised.getPosition() < 0) {
+            continue;
+          }
+          delta.setOriginal(
+              new Chunk<>(original.getPosition() + offset, original.getLines()));
+          delta.setRevised(
+              new Chunk<>(revised.getPosition() + offset, revised.getLines()));
+          PatchFailedException exception = applyDelta(delta, result);
+          if (exception == null) {
+            return;
+          }
+        }
+      }
+
+      throw e;
+    }
+
+    private static PatchFailedException applyDelta(Delta<String> delta, List<String> result) {
+      try {
+        delta.applyTo(result);
+        return null;
+      } catch (PatchFailedException e) {
+        return new PatchFailedException(e.getMessage() +
+            "\nFailed to apply delta:\n    " + delta.toString());
+      }
+    }
+  }
+
+  private enum LineType {
+    OLD_FILE,
+    NEW_FILE,
+    CHUNK_HEAD,
+    CHUNK_ADD,
+    CHUNK_DEL,
+    CHUNK_EQL,
+    GIT_HEADER,
+    RENAME_FROM,
+    RENAME_TO,
+    GIT_LINE,
+    UNKNOWN
+  }
+
+  private static final String[] GIT_LINE_PREFIXES = {
+      "old mode ",
+      "new mode ",
+      "deleted file mode ",
+      "new file mode ",
+      "copy from ",
+      "copy to ",
+      "rename old ",
+      "rename new ",
+      "similarity index ",
+      "dissimilarity index ",
+      "index "
+  };
+
+  private static LineType getLineType(String line, boolean isReadingChunk) {
+    if (isReadingChunk) {
+      if (line.startsWith("+")) {
+        return LineType.CHUNK_ADD;
+      }
+      if (line.startsWith("-")) {
+        return LineType.CHUNK_DEL;
+      }
+      if (line.startsWith(" ") || line.isEmpty()) {
+        return LineType.CHUNK_EQL;
+      }
+    } else {
+      if (line.startsWith("--- ")) {
```

Comment:
Should  be "--- " then, because in the other place we are doing substring(4).
Also below

Reply:
Done.